retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt

```diff
@@ -0,0 +1,225 @@
+            setAccessToken(null)
+          }
+          catch (exception: JacksonExceptions.JacksonParsingException)
+          {
+            warn("Call of refresh token failed with exception: ${exception.printStackTrace()}")
+            setAccessToken(null)
+          }
+
+          getAccessToken()?.apply {
+            val newAuthorization = "${this.tokenType} ${this.accessToken}"
+
+            // We don't want to try the call with the same token twice
+            if (response.request().header("Authorization") != newAuthorization)
+            {
+              return response.request().newBuilder()
+                  .header("Authorization", newAuthorization)
+                  .build()
+            }
+          }
+        }
+
+        // Fail case
+        setAccessToken(null)
+        return null
+      }
+    }
+
+    override fun intercept(chain: Interceptor.Chain): Response
+    {
+      val newRequest = chain.request().newBuilder()
+
+      authProvider.apply {
+        getXApiKey().takeIf { XApiKey -> XApiKey.isNotBlank() }?.apply {
+          newRequest.header("XApiKey", this)
+        }
+
+        getAccessToken()?.takeIf { accessToken -> accessToken.tokenType.isNotBlank() && accessToken.accessToken.isNotBlank() }?.apply {
+          newRequest.header("Authorization", "$tokenType $accessToken")
+        }
+      }
+
+      return chain.proceed(newRequest.build())
+    }
+
+    private fun responseCount(response: Response): Int
+    {
+      var responseHolder: Response? = response
+      var result = 1
+
+      while (result <= MAX_RETRIES && responseHolder?.priorResponse() != null)
+      {
+        result++
+        responseHolder = response.priorResponse()
+      }
+
+      return result
+    }
+  }
+
+  private val httpAuthClient: OkHttpClient by lazy {
+    OkHttpClient.Builder()
+        .connectTimeout(connectTimeout, TimeUnit.MILLISECONDS)
+        .readTimeout(readTimeout, TimeUnit.MILLISECONDS)
+        .writeTimeout(writeTimeout, TimeUnit.MILLISECONDS)
+        .addNetworkInterceptor(TokenAuthenticatorInterceptor())
+        .build()
+  }
+
+  private val authService: AuthAPI? by lazy {
+    authProvider.run {
+      val serviceBuilder = Retrofit
+          .Builder()
+          .baseUrl(getAuthRoute())
+          .client(httpAuthClient)
+
+      converterFactories.forEach { converterFactory ->
+        serviceBuilder.addConverterFactory(converterFactory)
+      }
+
+      return@run serviceBuilder.build().create(AuthAPI::class.java)
+    }
+  }
+
+
+  /**
+   * Method to login the user with the implemented authenticator.
+   *
+   * @param[username] the username of the user to connect.
+   * @param[password] the password of the user to connect.
+   *
+   * @return the success of the connection.
+   */
+  @WorkerThread
+  protected fun loginUser(username: String, password: String): ResponseWithError<AccessToken, ErrorResponse>?
```

Comment:
Here is a smaller function body doing the same thing : 
 
```suggestion
return try
    {
      executeAuth(authService?.authToken(authProvider.getAuthRoute(), username, password))
    }
    catch (exception: IOException)
    {
      warn("Call of login failed with exception: ${exception.printStackTrace()}")
      null
    }
    catch (exception: JacksonExceptions.JacksonParsingException)
    {
      warn("Call of login failed with exception: ${exception.printStackTrace()}")
      null
    }.also {
      authProvider.setAccessToken(it?.successResponse)
    }
```

Reply:
Done